prov/usnic/src/usdf_fabric.c

```diff
@@ -341,6 +341,94 @@ static int validate_modebits(uint32_t version, const struct fi_info *hints,
 	return FI_SUCCESS;
 }
 
+static int usdf_alloc_fid_nic(struct fi_info *fi,
+			struct usd_device_attrs *dap)
+{
+	int ret;
+	struct fid_nic *nic = NULL;
+	struct fi_device_attr *da = NULL;
+	struct fi_bus_attr *ba = NULL;
+	struct fi_link_attr *la = NULL;
+
+	nic = calloc(1, sizeof(*nic));
+	if (!nic)
+		goto nomem;
+	nic->fid.fclass = FI_CLASS_NIC;
+
+	da = calloc(1, sizeof(*da));
+	if (!da)
+		goto nomem;
+	nic->device_attr = da;
+	da->name = strdup(dap->uda_devname);
+	ret = asprintf(&da->device_id, "%s (%s)",
+		usd_devid_to_pid(dap->uda_vendor_id,
+				dap->uda_device_id),
+		usd_devid_to_nicname(dap->uda_vendor_id,
+				dap->uda_device_id));
+	if (ret < 0) {
+		USDF_DBG(
+			"asprintf failed while creating device_id\n");
+		ret = -ENOMEM;
+	}
+	ret = asprintf(&da->device_version, "0x%x", dap->uda_vendor_part_id);
+	if (ret < 0) {
+		USDF_DBG(
+			"asprintf failed while creating device_version\n");
+		ret = -ENOMEM;
+	}
+	ret = asprintf(&da->vendor_id, "0x%x", dap->uda_vendor_id);
+	if (ret < 0) {
+		USDF_DBG(
+			"asprintf failed while creating vendor_id\n");
+		ret = -ENOMEM;
+	}
+	da->driver = strdup("usnic_verbs");
+	da->firmware = strdup(dap->uda_firmware);
+
+	ba = calloc(1, sizeof(*ba));
+	if (!ba)
+		goto nomem;
+	// usnic does not currently expose PCI bus information, so we
+	// set the bus type to unknown.
+	ba->bus_type = FI_BUS_UNKNOWN;
+	nic->bus_attr = ba;
+
+	la = calloc(1, sizeof(*la));
+	if (!la)
+		goto nomem;
+	uint8_t *p = (uint8_t*) &dap->uda_ipaddr_be;
+	ret = asprintf(&la->address, "%d.%d.%d.%d/%d",
```

Comment:
Not that we use anything other than v4, but if we ever do support anything else this conversion can probably be replaced with `inet_ntop`.